### DISCUSSION TITLE
`check_next` question

### DISCUSSION THREAD

**Discussion by ghost** [savonet/liquidsoap#3816] [Q&A] [answered]
Hi! I've got a question and would like to ask for your help.

I know that it's possible to avoid repetitions of the file with this function:

```
l = playlog(persistency='p.log", duration=7200.)
def check(r)
m = request.metadata(r)
if l.last(m) < 3600. then
log.info("Rejecting #{m['filename']} (played #{l.last(m)}s ago).")
false
else
l.add(m)
true
end
end
s = playlist(check_next=check, "playlist")
output(s)
```

But what should be done/changed in order to check the Artist? I mean with the above function `file` will not repeat within an hour. And I want for the Artist to no repeat within an hour even if it's different songs from the same Artist. Hope I'm making myself clear.

Maybe it's possible to achieve somehow with making a file with playback history and checking through it? I'm not good at writing function stuff. So here I am.

Thank you!

**Comment by vitoyucepi** [comment]
Hi @Marco8585,
> But what should be done/changed in order to check the Artist?

That's pretty easy, you should use artist as the identifier.
`playlog(persistency="p.log", duration=7200., hash=metadata.artist)`
The `hash` argument accepts functions that take the metadata as an argument and return a string.
```ruby
def my_hash(m)
  metadata.artist(m)
end
```

**Reply by ghost** [reply]
Thanks! It worked. And I have to use new `playlog` function for every playlist? 

Also I have noticed that some artists appear in the `p.log` with spaces -> `Artist `. Maybe it is better to trim artist_name in case same artist would exist but without spaces in the name (`Artist ` vs `Artist`)? Could this be done and do I need it so checking would be robust?

Is there a "best practice" for such case in general (I mean not repeating an artist) taking into consideration my lack of LS skills? Something robust?

**Reply by vitoyucepi** [reply, marked as answer]
> And I have to use new playlog function for every playlist?

It depends on the desired behavior. Maybe you could use an already defined `check` function in each of them. You could also define your own function with the playlist encapsulated in it.

> Maybe it is better to trim artist_name in case same artist would exist but without spaces

Sure, yes
```ruby
def my_hash(m)
  artist = metadata.artist(m)
  string.trim(artist)
end

playlog(persistency="p.log", duration=7200., hash=my_hash)
```

> Is there a "best practice" for such case in general (I mean not repeating an artist) taking into consideration my lack of LS skills? Something robust?

The `playlog` function is a good solution, but you may want to create your own scheduler. For example, azuracast relies on it's own autodj feature implemented with php.

**Reply by ghost** [reply]
I understand. One more question if I may. Is it possible to set `my_hash(m)` function or do anything else to handle following issue:

Some artists are represented with different names, e.g.
Artist name: `Bob Marley`
Artist name: `Bob Marley and the Wailers`
Artist name: `Bob Marley, The Wailers`
Artist name: `Bob Marley & the Wailers`
Artist name: `Bob Marley with the Wailers`

But nevertheless it's the same artist. Is there a way to consider this? Maybe "cut" an artist and divide it or use a name before `,/&/and/with` somehow?

**Reply by vitoyucepi** [reply]
Perhaps a good solution would be
1. extract words into a list;
2. normalize, lower case and sort the words;
3. remove stop words;
4. use the rest as a key.

**Reply by ghost** [reply]
Wouldn't know how to do it. But thanks nevertheless for helpful answers.

**Reply by vitoyucepi** [reply]
Maybe something like this
<details>

```ruby
stopwords=[
"a",
"about",
"above",
"after",
"again",
"against",
"all",
"am",
"an",
"and",
"any",
"are",
"aren't",
"as",
"at",
"be",
"because",
"been",
"before",
"being",
"below",
"between",
"both",
"but",
"by",
"can't",
"cannot",
"could",
"couldn't",
"did",
"didn't",
"do",
"does",
"doesn't",
"doing",
"don't",
"down",
"during",
"each",
"few",
"for",
"from",
"further",
"had",
"hadn't",
"has",
"hasn't",
"have",
"haven't",
"having",
"he",
"he'd",
"he'll",
"he's",
"her",
"here",
"here's",
"hers",
"herself",
"him",
"himself",
"his",
"how",
"how's",
"i",
"i'd",
"i'll",
"i'm",
"i've",
"if",
"in",
"into",
"is",
"isn't",
"it",
"it's",
"its",
"itself",
"let's",
"me",
"more",
"most",
"mustn't",
"my",
"myself",
"no",
"nor",
"not",
"of",
"off",
"on",
"once",
"only",
"or",
"other",
"ought",
"our",
"ours",
"ourselves",
"out",
"over",
"own",
"same",
"shan't",
"she",
"she'd",
"she'll",
"she's",
"should",
"shouldn't",
"so",
"some",
"such",
"than",
"that",
"that's",
"the",
"their",
"theirs",
"them",
"themselves",
"then",
"there",
"there's",
"these",
"they",
"they'd",
"they'll",
"they're",
"they've",
"this",
"those",
"through",
"to",
"too",
"under",
"until",
"up",
"very",
"was",
"wasn't",
"we",
"we'd",
"we'll",
"we're",
"we've",
"were",
"weren't",
"what",
"what's",
"when",
"when's",
"where",
"where's",
"which",
"while",
"who",
"who's",
"whom",
"why",
"why's",
"with",
"won't",
"would",
"wouldn't",
"you",
"you'd",
"you'll",
"you're",
"you've",
"your",
"yours",
"yourself",
"yourselves",
]

def tokenize(a)
  b = string.case(a, lower=true)
  c = r/[^\w'-]+/.split(b)
  # TODO: normalize
  d = list.filter(fun(i) -> not list.mem(i, stopwords), c)
  list.sort.natural(d)
end

def my_hash(m)
  artist = metadata.artist(m)
  string.concat(tokenize(artist), separator=" ")
end
```

</details>

**Reply by ghost** [reply]
Thanks so much!

**Reply by ghost** [reply]
Hi again, @vitoyucepi. I would like to ask followup question. 

In `stopwords=[]` I would like to use `&` becasue it's very common and often `&` and `and` change each other in the same artist.

But when I add `&` to `stopwords=[]` it always results in additional space in front. Example:

`Bob Marley & the Wailers`
After function:
`  bob marley wailers` - double space in front. So `&` is represented with ` ` then another ` ` separator.

What can be done to avoid it?

**Reply by vitoyucepi** [reply]
Oh, I'm sorry. I made a mistake.
```ruby
c = r/[^\w'-]+/.split(b)
```

**Reply by ghost** [reply]
Thanks. It works! But another issue I've noticed. If I have something with `ó` o-acute -> `revolución` becomes `n revoluci`. Same with other non-purely-english symbols. In "getting job done" context it's probably ok.
Could anything be done to avoid it or maybe it's alright?

**Reply by vitoyucepi** [reply]
It's quite a difficult task to determine word boundaries. You could stuff the regex with something like `\PL` or `\pZ\pP\pS`, but the result is not very predictable. Maybe you could send your string to a separate program and it will return the correct words.
Here are some libs that might be useful in this situation
- https://lucene.apache.org
- https://crates.io/crates/unicode-segmentation

**Reply by ghost** [reply]
Thanks @vitoyucepi. I'll try and make something up.

**Comment by Moonbase59** [comment]
It all very much depends upon how good your files are tagged.

I’d probably use the _[MusicBrainz Artist ID](https://musicbrainz.org/doc/MusicBrainz_Identifier)_ to identify an artist (or maybe even the _MusicBrainz Artist Sortname_), and the [ISRC](https://musicbrainz.org/doc/ISRC) (if known) for song identification.

But @vitoyucepi’s script examples—as usual—are great!
